src/util/result.h

```diff
@@ -9,41 +9,60 @@
 
 #include <variant>
 
-/*
- * 'BResult' is a generic class useful for wrapping a return object
- * (in case of success) or propagating the error cause.
-*/
+namespace util {
+
+struct Error { bilingual_str message; };
+
+//! The util::Result class provides a standard way for functions to return error
+//! messages in addition to optional result values.
+//!
+//! It is intended for high-level functions that need to report error strings to
+//! end users. Lower-level functions that don't need this error-reporting and
+//! only need error-handling should avoid util::Result and instead use standard
+//! classes like std::optional, std::variant, and std::tuple, or custom structs
+//! and enum types to return function results.
+//!
+//! Usage examples can be found in \example ../test/result_tests.cpp, but in
+//! general code returning `util::Result<T>` values is very similar to code
+//! returning `std::optional<T>` values. Existing functions returning
+//! `std::optional<T>` can be updated to return `util::Result<T>` and return
+//! error strings usually just replacing `return std::nullopt;` with `return
+//! util::Error{error_string};`.
 template<class T>
-class BResult {
+class Result {
 private:
     std::variant<bilingual_str, T> m_variant;
 
+    template <typename FT>
+    friend class Result;
+    template <typename FT>
+    friend bilingual_str ErrorString(const Result<FT>& result);
+
 public:
-    BResult() : m_variant{Untranslated("")} {}
-    BResult(T obj) : m_variant{std::move(obj)} {}
-    BResult(bilingual_str error) : m_variant{std::move(error)} {}
-
-    /* Whether the function succeeded or not */
-    bool HasRes() const { return std::holds_alternative<T>(m_variant); }
-
-    /* In case of success, the result object */
-    const T& GetObj() const {
-        assert(HasRes());
-        return std::get<T>(m_variant);
-    }
-    T ReleaseObj()
-    {
-        assert(HasRes());
-        return std::move(std::get<T>(m_variant));
-    }
-
-    /* In case of failure, the error cause */
-    const bilingual_str& GetError() const {
-        assert(!HasRes());
-        return std::get<bilingual_str>(m_variant);
-    }
-
-    explicit operator bool() const { return HasRes(); }
+    Result(T obj) : m_variant{std::move(obj)} {}
+    Result(Error error) : m_variant{std::move(error.message)} {}
+    template <typename OT>
+    Result(Error error, Result<OT>&& other) : m_variant{std::move(std::get<0>(other.m_variant))} {}
+
+    //! std::optional methods, so functions returning optional<T> can change to
+    //! return Result<T> with minimal changes to existing code, and vice versa.
+    bool has_value() const { return m_variant.index() == 1; }
+    const T& value() const { assert(*this); return std::get<1>(m_variant); }
+    T& value() { assert(*this); return std::get<1>(m_variant); }
+    template <typename U>
+    T value_or(const U& default_value) const { return has_value() ? value() : default_value; }
```

Comment:
This doesn't compile, so my recommendation would be to remove dead code and only add it when it is needed.


```
./util/result.h:53:69: error: call to implicitly-deleted copy constructor of 'const result_tests::NoCopy'
    T value_or(const U& default_value) const { return has_value() ? value() : default_value; }
                                                                    ^~~~~~~
test/result_tests.cpp:74:27: note: in instantiation of function template specialization 'util::Result<result_tests::NoCopy>::value_or<int>' requested here
    BOOST_CHECK_EQUAL(obj.value_or(1), 5);
                          ^
test/result_tests.cpp:14:26: note: copy constructor of 'NoCopy' is implicitly deleted because field 'm_n' has a deleted copy constructor
    std::unique_ptr<int> m_n;
                         ^
```

```diff
diff --git a/src/test/result_tests.cpp b/src/test/result_tests.cpp
index 5118571e23..bfd5672045 100644
--- a/src/test/result_tests.cpp
+++ b/src/test/result_tests.cpp
@@ -68,4 +68,11 @@ BOOST_AUTO_TEST_CASE(check_returned)
     ExpectFail(NoCopyFn(5, false), Untranslated("nocopy 5 error."));
 }
 
+BOOST_AUTO_TEST_CASE(check_value_or)
+{
+    auto obj{NoCopyFn(5, true)};
+    BOOST_CHECK_EQUAL(obj.value_or(1), 5);
+    obj=NoCopyFn(5, false);
+    BOOST_CHECK_EQUAL(obj.value_or(1), 1);
+}
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/util/result.h b/src/util/result.h
index b9e490025b..71aec57237 100644
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -49,8 +49,6 @@ public:
     bool has_value() const { return m_variant.index() == 1; }
     const T& value() const { assert(*this); return std::get<1>(m_variant); }
     T& value() { assert(*this); return std::get<1>(m_variant); }
-    template <typename U>
-    T value_or(const U& default_value) const { return has_value() ? value() : default_value; }
     operator bool() const { return has_value(); }
     const T* operator->() const { return &value(); }
     const T& operator*() const { return value(); }